crates/runtime/src/memfd.rs

```diff
@@ -0,0 +1,566 @@
+        {
+            self.dirty = true;
+            self.cur_size = initial_size_bytes;
+            return Ok(());
+        }
+
+        // Otherwise, we need to redo (i) the anonymous-mmap backing
+        // for the whole slot, (ii) the initial-heap-image mapping if
+        // present, and (iii) the mprotect(PROT_NONE) above the
+        // initial heap size.
+
+        // Security/audit note: we map all of these MAP_PRIVATE, so
+        // all instance data is local to the mapping, not propagated
+        // to the backing fd. We throw away this CoW overlay with
+        // madvise() below, from base up to static_size (which is the
+        // whole slot) when terminating the instance.
+
+        // Anonymous mapping behind the initial heap size: this gives
+        // zeroes for any "holes" in the initial heap image. Anonymous
+        // mmap memory is faster to fault in than a CoW of a file,
+        // even a file with zero holes, because the kernel's CoW path
+        // unconditionally copies *something* (even if just a page of
+        // zeroes). Anonymous zero pages are fast: the kernel
+        // pre-zeroes them, and even if it runs out of those, a memset
+        // is half as expensive as a memcpy (only writes, no reads).
+        //
+        // We map these inaccessible at first then mprotect() the
+        // whole of the initial heap size to R+W below.
+        //
+        // Special case: we can skip if the last instantiation had no
+        // image. This means that the whole slot is filled with an
+        // anonymous mmap backing (and it will have already been
+        // cleared by the madvise). We may however need to
+        // mprotect(NONE) the space above `initial_size_bytes` if the
+        // last use of this slot left it larger. This also lets us
+        // skip an mmap the first time a MemFdSlot is used, because we
+        // require the caller to give us a fixed address in an
+        // already-mmaped-with-anon-memory region. This is important
+        // for the on-demand allocator.
+        if self.image.is_some() {
+            self.map_anon_memory(rustix::io::ProtFlags::empty())
+                .map_err(|e| InstantiationError::Resource(e.into()))?;
+        } else if initial_size_bytes < self.initial_size {
+            self.set_protection(
+                initial_size_bytes,
+                self.initial_size,
+                rustix::io::MprotectFlags::empty(),
+            )
+            .map_err(|e| InstantiationError::Resource(e.into()))?;
+        }
+
+        // The initial memory image, if given. If not, we just get a
+        // memory filled with zeroes.
+        if let Some(image) = maybe_image {
+            if image.len > 0 {
+                let image = image.clone();
+
+                unsafe {
+                    let ptr = rustix::io::mmap(
+                        (self.base + image.offset) as *mut c_void,
+                        image.len,
```

Comment:
Now that I think about it, mind adding some debug asserts here that `image.offset + image.len` fits within `initial_size_bytes`?

Reply:
Done! (And I made it a release assert rather than debug-only, as this feels important enough and we use release asserts elsewhere in this file)